util/stream_dataset.py

```diff
@@ -471,8 +483,12 @@ def _interpolate_and_import_needed(self, param, external_datasets):
                 source, value = funcmap[name]
                 if source not in ['CAL', self.stream_key.stream]:
                     source_key = streams.get(source)
-                    if source_key in external_datasets:
-                        self.interpolate_into(source_key, external_datasets[source_key], value)
+                    # prevent trying to interpolate with unpopulated virtual streams
+                    # if interpolating virtual streams, skip other parameters
+                    if (interpolate_virtual and source_key.is_virtual) or not (
```

Comment:
Is this saying "if both flags are virtual or both flags are not virtual"?

Reply:
Yes. In other words, we want to interpolate virtual streams and the stream key indicates a virtual stream, OR, we do not want to interpolate virtual streams and the stream key does not indicate virtual. I just refactored "(A and B) or (~A and ~B)" to be "(A and B) or ~(A or B)".

Reply:
This code slightly confusing because "interpolate_virtual" is somewhat vaguely named. If interpolate_virtual is true, it means we want to interpolate only virtual streams; if interpolate_virtual is false, it means we only want to interpolate non-virtual streams. In Java, we would use an enum to clarify this. Can something similar be done in Python?

Reply:
I hear what you're saying, but trying to make an enum (or equivalent) here seems like overkill to me, especially with the updated docstring.